Tests/UnitTests/Core/EventData/MultiTrajectoryTests.cpp

```diff
@@ -520,98 +541,87 @@ BOOST_AUTO_TEST_CASE(trackstate_reassignment) {
-        BOOST_CHECK_EQUAL(tsProxy.effectiveProjector(), meas.projector());
-      },
-      *ts.measurement.calibrated);
-}
+  BOOST_CHECK_EQUAL(&ts.referenceSurface(), &pc.sl.referenceSurface());
 
-BOOST_AUTO_TEST_CASE(add_trackstate_allocations) {
-  auto [ts, fm] = make_trackstate();
+  BOOST_CHECK(ts.hasJacobian());
+  BOOST_CHECK_EQUAL(ts.jacobian(), pc.jacobian);
 
-  ts.parameter.filtered = std::nullopt;
-  ts.parameter.smoothed = std::nullopt;
-  ts.measurement.calibrated = std::nullopt;
+  BOOST_CHECK(ts.hasProjector());
+  BOOST_CHECK_EQUAL(ts.effectiveProjector(), pc.meas3d->projector());
+  // measurement properties
+  BOOST_CHECK(ts.hasCalibrated());
+  BOOST_CHECK_EQUAL(pc.meas3d->parameters(), ts.effectiveCalibrated());
+  ParVec_t mParFull;
+  mParFull.setZero();
+  mParFull.head(pc.meas3d->size()) = pc.meas3d->parameters();
+  BOOST_CHECK_EQUAL(mParFull, ts.calibrated());
 
-  MultiTrajectory<SourceLink> mj;
+  BOOST_CHECK_EQUAL(pc.meas3d->covariance(),
+                    ts.effectiveCalibratedCovariance());
+  CovMat_t mCovFull;
+  mCovFull.setZero();
+  mCovFull.topLeftCorner(pc.meas3d->size(), pc.meas3d->size()) =
+      pc.meas3d->covariance();
+  BOOST_CHECK_EQUAL(mCovFull, ts.calibratedCovariance());
 
-  // this should allocate for all the components in the trackstate, plus
-  // filtered
-  size_t i = mj.addTrackState(ts, TrackStatePropMask::Filtered);
+  // calibrated links to original measurement
+  BOOST_CHECK_EQUAL(pc.meas3d->sourceLink(), ts.calibratedSourceLink());
 
-  auto tsp = mj.getTrackState(i);
+  // uncalibrated **is** a SourceLink
+  BOOST_CHECK(ts.hasUncalibrated());
+  BOOST_CHECK_EQUAL(pc.meas3d->sourceLink(), ts.uncalibrated());
+
+  // full projector, should be exactly equal
+  CovMat_t fullProj;
+  fullProj.setZero();
+  fullProj.topLeftCorner(pc.meas3d->size(),
+                         MultiTrajectory<SourceLink>::MeasurementSizeMax) =
+      pc.meas3d->projector();
+  BOOST_CHECK_EQUAL(ts.projector(), fullProj);
+
+  // projector with dynamic rows
+  // should be exactly equal
+  BOOST_CHECK_EQUAL(ts.effectiveProjector(), pc.meas3d->projector());
+}
 
-  BOOST_CHECK(tsp.hasPredicted());
-  BOOST_CHECK(tsp.hasFiltered());
-  BOOST_CHECK(!tsp.hasSmoothed());
-  BOOST_CHECK(tsp.hasUncalibrated());
-  BOOST_CHECK(!tsp.hasCalibrated());
-  BOOST_CHECK(tsp.hasJacobian());
+BOOST_AUTO_TEST_CASE(add_trackstate_allocations) {
+  MultiTrajectory<SourceLink> t;
+
+  // this should allocate for all the components in the trackstate, plus
+  // filtered
+  size_t i = t.addTrackState(
+      TrackStatePropMask::Predicted | TrackStatePropMask::Filtered |
+      TrackStatePropMask::Uncalibrated | TrackStatePropMask::Jacobian);
+  auto tso = t.getTrackState(i);
+  make_trackstate(tso, TrackStatePropMask::Predicted);
+  make_trackstate(tso, TrackStatePropMask::Filtered);
+  make_trackstate(tso, TrackStatePropMask::Uncalibrated);
+  make_trackstate(tso, TrackStatePropMask::Jacobian);
```

Comment:
Ok I see if you only ever call it like this it's fine.

Reply:
Yeah, I had to adapt the logic a bit